File: BOJ_Python/자료구조_스택/BOJ_9012_괄호.py
import sys

n = int(sys.stdin.readline())

for i in range(n) :
    stack = []
    s = sys.stdin.readline()
    for i in s :
        if i == "(":
            stack.append(i)
        elif i == ")" :
            if len(stack) == 0:
                print("NO")
                break
            else: 
                stack.pop()

    else :
        if len(stack) == 0:
            print("YES")
        else:
            print("NO")

File: BOJ_Swift/스택/BOJ_9012_괄호.swift
let n = Int(readLine()!)!

for _ in 0..<n {
    var stack = []
    let s = readLine()!
    for a in s {
        if a == "(" {
            stack.append(a)
        } 
        else if String(a) == ")" {
            if len(stack) == 0 {
                print("NO")
                break
            } else {
                stack.pop()
            }
        }
    }
    if len(stack) == 0 {
        print("YES")
    } else {
        print("NO")
    }
}
